File: 0149-max-points-on-a-line/0149-max-points-on-a-line.cpp
class Solution {
private:
    int bruteForce(vector<vector<int>>& points){
        
        int n = points.size();
        if(n == 1) return 1;
        
        int maxPoint = 0;
        
        for(int i=0; i<n; i++){ // p1 (x1, y1)
            
            for(int j=i+1; j<n; j++){ // p2(x2, y2)
                
                int count = 2;
                
                int dx = points[j][0] - points[i][0]; // (x2-x1)
                int dy = points[j][1] - points[i][1]; // (y2-y1)
                
                /*
                    dy/dx = dy_/dx_
                    dy*dx_ = dy_*dx
                */
                
                for(int k=0; k<n; k++) { // p3(x3, y3)
                    
                    if(k != i && k != j){
                        
                        int dx_ = points[k][0] - points[i][0]; // (x3-x1)
                        int dy_ = points[k][1] - points[i][1]; // (y3-y1)
                        
                        if(dy*dx_ == dy_*dx) count++;
                    }
                }
                
                maxPoint = max(maxPoint, count);
            }
        }
        
        return maxPoint;
    }
    
public:
    int maxPoints(vector<vector<int>>& points) {
        return bruteForce(points);
    }
};

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.js
/**
 * @param {number[][]} points
 * @return {number}
 */
function findGCD(a, b) {
    
    if( b == 0) return a;
    
    return findGCD(b, a % b);
}

var maxPoints = function(points) {
    
    let n = points.length;
    
    if(n == 1)
        return 1;
    
    let result = 0;
    
    for(let i=0; i<n; i++){
        
        let freq = new Map();
        
        for(let j=0; j<n; j++){
            
            if(i == j) continue;
            
            let dx = points[j][0] - points[i][0];
            let dy = points[j][1] - points[i][1];
            
            let gcd = findGCD(Math.max(dx, dy), Math.min(dx, dy));
            
            let key = (dx / gcd) + "_" + (dy / gcd);
            
            if(freq.has(key) == false){
                freq.set(key, 1);
            }
            else{
                freq.set(key, freq.get(key) + 1);
            }
        }
        
        freq.forEach((value, key) => {
            result = Math.max(result, value + 1);
        })
    }
    
    return result;
};